utils/runtime-fuzzer/src/generator.rs

```diff
@@ -0,0 +1,259 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+mod claim_value;
+mod send_message;
+mod send_reply;
+mod upload_program;
+
+use gear_call_gen::GearCall;
+use gear_common::event::ProgramChangeKind;
+use gear_core::ids::{MessageId, ProgramId};
+use gear_wasm_gen::wasm_gen_arbitrary::{Result, Unstructured};
+use pallet_balances::Pallet as BalancesPallet;
+use pallet_gear::Event as GearEvent;
+use runtime_primitives::AccountId;
+use std::{collections::BTreeSet, mem};
+use vara_runtime::{Runtime, RuntimeEvent, System};
+
+use crate::{data::*, runtime};
+
+// Max code size - 25 KiB.
+const MAX_CODE_SIZE: usize = 25 * 1024;
+
+/// Maximum payload size for the fuzzer - 1 KiB.
+///
+/// TODO: #3442
+const MAX_PAYLOAD_SIZE: usize = 1024;
+const _: () = assert!(MAX_PAYLOAD_SIZE <= gear_core::message::MAX_PAYLOAD_SIZE);
+
+/// Maximum salt size for the fuzzer - 512 bytes.
+///
+/// There's no need in large salts as we have only 35 extrinsics
+/// for one run. Also small salt will make overall size of the
+/// corpus smaller.
+const MAX_SALT_SIZE: usize = 512;
+const _: () = assert!(MAX_SALT_SIZE <= gear_core::message::MAX_PAYLOAD_SIZE);
+
+const ID_SIZE: usize = mem::size_of::<ProgramId>();
+const GAS_AND_VALUE_SIZE: usize = mem::size_of::<(u64, u128)>();
+
+/// Used to make sure that generators will not exceed `Unstructured` size as it's used not only
+/// to generate things like wasm code or message payload but also to generate some auxiliary
+/// data, for example index in some vec.
+const AUXILIARY_SIZE: usize = 512;
+
+pub(crate) struct GearCallsGenerator<'a> {
```

Comment:
I don't agree with your `TempData` implementation. It stored program ids of non existent programs. The only way to get real data from runtime is to do it with events. I didn't see how it can be included into your `Iterator` implementation for `GearCalls`.

Reply:
the idea with an array of calls seems interesting, let me try

Reply:
discussed.

The idea with an array is great, but has a problem that `inp.try_into()?` returns a different error, not `arbitrary::Error`.